latch_cli/services/cp/download.py

```diff
@@ -0,0 +1,176 @@
+from latch_cli import tinyrequests
+from latch_cli.config.latch import config as latch_config
+from latch_cli.services.cp.config import CPConfig, Progress
+from latch_cli.services.cp.node_utils import LDataNodeType, get_node_data
+from latch_cli.services.cp.path_utils import normalize_path
+from latch_cli.services.cp.progress import ProgressBarManager, ProgressBars
+from latch_cli.services.cp.utils import get_auth_header, human_readable_time, pluralize
+from latch_cli.utils import with_si_suffix
+
+
+class GetSignedUrlData(TypedDict):
+    url: str
+
+
+class GetSignedUrlsRecursiveData(TypedDict):
+    urls: Dict[str, str]
+
+
+def download(
+    src: str,
+    dest: Path,
+    config: CPConfig,
+):
+    normalized = normalize_path(src)
+    node_data = get_node_data(normalized)
+
+    node_has_children = node_data.type in {
+        LDataNodeType.account_root,
+        LDataNodeType.dir,
+        LDataNodeType.mount,
+    }
+
+    if node_has_children:
+        endpoint = latch_config.api.data.get_signed_urls_recursive
+    else:
+        endpoint = latch_config.api.data.get_signed_url
+
+    res = tinyrequests.post(
+        endpoint,
+        headers=get_auth_header(),
+        json={"path": normalized},
+    )
+
+    if res.status_code != 200:
+        raise ValueError(f"Failed to download {src}: {res.json()['error']}")
+
+    json_data = res.json()
+    if node_has_children:
+        dir_data: GetSignedUrlsRecursiveData = json_data["data"]
+
+        if dest.exists() and not dest.is_dir():
+            raise ValueError(f"{str(dest)} is not a directory")
+
+        if dest.exists() and not normalized.endswith("/"):
+            dest = dest / node_data.name
+
+        urls: List[str] = []
+        dests: List[Path] = []
+
+        for rel_path, url in dir_data["urls"].items():
+            urls.append(url)
+            dests.append(dest / rel_path)
+
+        num_files = len(urls)
+
+        if config.progress == Progress.none:
+            num_bars = 0
+            show_total_progress = False
+        if config.progress == Progress.total:
+            num_bars = 0
+            show_total_progress = True
+        else:
+            num_bars = min(config.max_concurrent_files, num_files)
+            show_total_progress = True
+
+        with ProgressBarManager() as manager:
+            progress_bars: ProgressBars = manager.ProgressBars(
+                num_bars,
+                show_total_progress=show_total_progress,
+                verbose=config.verbose,
+            )
+            progress_bars.set_total(num_files, "Copying Files")
+
+            start = time.perf_counter()
+            with ProcessPoolExecutor() as executor:
+                total_bytes = sum(
+                    executor.map(
+                        download_file,
+                        urls,
+                        dests,
+                        repeat(config.chunk_size),
+                        repeat(progress_bars),
+                    )
+                )
```

Comment:
Do you have an idea of the difference in performance when using processes vs an async loop?

Reply:
added a todo to benchmark